### PR TITLE
feat: consider various lengths for dynamic-sized parameters

### PR DESCRIPTION
**Pull request by daejunpark**
Previously, only a single length for dynamic-sized parameters in test functions was considered.

Now, users can specify multiple lengths, and all possible combinations of these lengths will be considered during symbolic execution. Also, users can now set default lengths for dynamic arrays independently of the loop unrolling bound, whereas they were previously tied to the --loop value implicitly.

Flags to specify lengths:
- --array-lengths: specifies lengths for dynamic parameters (including dynamic arrays, bytes, and strings)
  - syntax: `name1={size1,size2,...},name2=...,` #283
- --default-array-lengths: set default lengths for dynamic arrays (default: `0,1,2`)
- --default-bytes-lengths: set default lengths for bytes and strings (default: `0,65,1024`)

### CODE REVIEW

File: src/halmos/config.py [hunk 1]
```diff
@@ -109,18 +110,29 @@ def parse(values: str | None) -> dict[str, list[int]]:
         if not values:
             return {}
 
-        # TODO: update syntax: name1={size1,size2},name2=size3,...
+        # syntax: --array-lengths name1=sizes1,name2=sizes2,...
+        # where sizes is either a comma-separated list of integers enclosed in curly braces, or a single integer
+
+        # remove all white spaces to simplify the pattern matching
+        values = "".join(values.split())
```
Comment (karmacoma-eth) [inline src/halmos/config.py:117]: I don't understand this line, we don't expect whitespace in the middle of the values, right? why not just `strip()`?

File: src/halmos/config.py [hunk 2]
```diff
@@ -231,15 +243,22 @@ class Config:
     )
 
     array_lengths: str = arg(
-        help="set the length of dynamic-sized arrays including bytes and string (default: loop unrolling bound)",
+        help="specify lengths for dynamic-sized arrays, bytes, and string. Lengths can be specified as a comma-separated list of integers enclosed in curly braces, or as a single integer.",
```
Comment (karmacoma-eth) [inline src/halmos/config.py:246]:
```suggestion
        help="specify lengths for dynamic-sized arrays, bytes, and string types. Lengths can be specified as a comma-separated list of integers enclosed in curly braces, or as a single integer.",
```

Comment (karmacoma-eth) [reply src/halmos/config.py:246]: does it really work with curly braces? in my shell it has a special meaning and does substitutions:

```sh
$ echo a={1,2}
a=1 a=2
```

File: src/halmos/config.py [hunk 3]
```diff
@@ -231,15 +243,22 @@ class Config:
...
+        help="set default lengths for dynamic-sized arrays (excluding bytes and string) not specified in --array-lengths",
+        global_default="0,1,2",
+        metavar="LENGTH1,LENGTH2,...",
+        action=ParseCSV,
+    )
+
     default_bytes_lengths: str = arg(
-        help="set the default length candidates for bytes and string not specified in --array-lengths",
-        global_default="0,1024,65",  # 65 is ECDSA signature size
+        help="set default lengths for bytes and string not specified in --array-lengths",
```
Comment (karmacoma-eth) [inline src/halmos/config.py:260]:
```suggestion
        help="set default lengths for bytes and string types not specified in --array-lengths",
```

**Review by karmacoma-eth — APPROVED** [review]

File: src/halmos/config.py (hunk 2 above)
Comment (daejunpark) [reply src/halmos/config.py:246]: um, you can escape curly braces in various ways, e.g., `echo 'a={1,2}'`

File: src/halmos/config.py (hunk 1 above)
Comment (daejunpark) [reply src/halmos/config.py:117]: we do expect whitespaces, like: `halmos --array-lengths "x={1,   2}"`

**Comment by daejunpark** [comment]
there's a multithreading-related issue with pytest in ci, which is orthogonal to this pr. merging this despite some test failures. trying to fix the flakiness separately.

**Pull request merged**